### PR TITLE
[CINN]apply broadcast tree optimization

### PR DESCRIPTION
**Pull request by Hongqing-work**
### PR Category

CINN

### PR Types

 Improvements

### Description

Pcard-67164
This PR applies the optimization of  broadcast tree lowering to host function.
To learn more about the theoretical explanation of lowering broadcast tree in host wrapper function and the step-by-step arrangement for the Pull Request (PR), please refer to the [draft PR65604](https://github.com/PaddlePaddle/Paddle/pull/65604) for details.

### CODE REVIEW

File: paddle/cinn/hlir/dialect/operator/transforms/lowering_pass/lower_cinn_fusion_op_pass.cc [hunk 1]
```diff
@@ -32,35 +32,31 @@ namespace cinn::dialect::ir::details {
...
+    const auto& optional_broadcast_tree = GetBroadcastTreeForOptimize(group);
+    if (optional_broadcast_tree.has_value()) {
+      const std::shared_ptr<BroadcastTree> broadcast_tree =
+          optional_broadcast_tree.value();
+      const auto& value_to_dim_expr_idx =
+          GetGroupDimExprInfo(group).value_to_dim_expr_idx;
+      return CompileBroadcastTree(
+          group, *broadcast_tree, value_to_dim_expr_idx);
+    } else {
+      return GetJitKernelAttr(group);
```
Comment (Hongqing-work) [inline paddle/cinn/hlir/dialect/operator/transforms/lowering_pass/lower_cinn_fusion_op_pass.cc:45]: 当前PR暂时还是将Broadcast tree的编译放在ProcessDyShapeGroup中，下一个PR将直接放入GetJitKernelAttr函数的逻辑中并放入PreAnalysis增强系统可维护性

File: paddle/cinn/hlir/dialect/operator/transforms/lowering_pass/broadcast_with_cf.h [hunk 2]
```diff
@@ -35,12 +35,8 @@ std::optional<std::shared_ptr<BroadcastTree>> GetBroadcastTreeForOptimize(
 bool ContainBroadcastShape(const common::BroadcastLeaf& leaves);
 GroupDimExprInfo GetGroupDimExprInfo(const OpLoweringGroupPtr& group);
 
-pir::Operation* CompileBroadcastTreeToConditionBlock(
+std::unordered_map<std::string, pir::Attribute> CompileBroadcastTree(
```
Comment (tc20042008) [inline paddle/cinn/hlir/dialect/operator/transforms/lowering_pass/broadcast_with_cf.h:38]: CompileBroadcastTreeToJitKernelOpAttrs
函数签名发生了变化，名字必须做到名副其实。

File: paddle/cinn/hlir/dialect/operator/transforms/lowering_pass/utils.cc [hunk 3]
```diff
@@ -58,24 +58,17 @@ std::vector<pir::Value> GetBlockOutsideInput(
   return vec_res;
 }
 
-std::unordered_map<OpLoweringGroupPtr,
-                   std::unordered_map<std::string, pir::Attribute>>
-CompileGroupAsOpAttribute(const std::vector<OpLoweringGroupPtr>& group_list) {
+std::unordered_map<std::string, pir::Attribute>
+CompileBroadcastGroupsAsOpAttribute(
```
Comment (tc20042008) [inline paddle/cinn/hlir/dialect/operator/transforms/lowering_pass/utils.cc:62]: CompileBroadcastGroupsAsJitKernelOpAttribute

语义要表达到位

File: paddle/cinn/hlir/framework/pir/op_lowering_group.cc [hunk 4]
```diff
@@ -130,32 +130,14 @@ void OpLoweringGroup::SetShapeOrDataExprs(
...
-    VLOG(4) << "clone op :" << op->name();
-    auto* new_op = op->Clone(*ir_mapping, clone_options);
-    // NOTE(dev): Must call block.insert to deal with ownership, otherwise it
-    // will lead memory-leak.
-    target_block->insert(target_block->end(), new_op);
-    new_ops.push_back(new_op);
-    ops_mapper[op] = new_op;
-  }
-
+    const int& group_idx) const {
```
Comment (tc20042008) [inline paddle/cinn/hlir/framework/pir/op_lowering_group.cc:133]: `const std::string& name_suffix`
这个字段的作用太弱了，其实本函数需要的只是名字后缀。到时候直接把`std::to_string(group_idx)`传给本函数就行。

**Review by tc20042008 — APPROVED** [review]

**Review by zyfncg — APPROVED** [review]

File: paddle/cinn/hlir/framework/pir/op_lowering_group.cc (hunk 4 above)
Comment (Hongqing-work) [reply paddle/cinn/hlir/framework/pir/op_lowering_group.cc:133]: 收到，上述命名及变量问题都将在下一个PR中进行修改

**Pull request merged**